Hook up storage for persistent registered content scripts.

kiaraarose opened the pull request:
#### 4485c484a89d836f7ddac832426ddfae1ea14fad
<pre>
Hook up storage for persistent registered content scripts.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265899">https://bugs.webkit.org/show_bug.cgi?id=265899</a>

Reviewed by Timothy Hatcher.

This is the final work for scripting API. This patch creates a _WKWebExtensionRegisteredScriptsSQLiteStore
which will be used to manage storage for registered scripts. This patch also loads these scripts
from storage when an extension loads.

When an extension updates, we will need to remove these scripts from storage since they shouldn&apos;t
persist in this case. This is being tracked in <a href="https://bugs.webkit.org/show_bug.cgi?id=249266.">https://bugs.webkit.org/show_bug.cgi?id=249266.</a>

* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.mm: Added.
(rowFilterStringFromRowKeys):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore initWithUniqueIdentifier:directory:usesInMemoryDatabase:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore updateScripts:completionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore deleteScriptsWithIDs:completionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore addScripts:completionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore getScriptsWithCompletionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _getScriptsWithOutErrorMessage:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _getKeysAndValuesFromRowEnumerator:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _insertScript:inDatabase:errorMessage:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _currentDatabaseSchemaVersion]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _databaseURL]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _createFreshDatabaseSchema]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _resetDatabaseSchema]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _isDatabaseEmpty]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingRegisterContentScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingGetRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterContentScripts):
(WebKit::WebExtensionContext::loadRegisteredContentScripts):
(WebKit::WebExtensionContext::createInjectedContentForScripts):
Renamed from parseRegisteredContentScripts. Parse scripts and create injected content data.
(WebKit::WebExtensionContext::parseRegisteredContentScripts): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::unload):
(WebKit::WebExtensionContext::invalidateStorage):
Invalidate all references to extension storage.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::registeredContentScriptsDirectory):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::parseRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::parseRegisteredContentScripts): Deleted.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:

Canonical link: <a href="https://commits.webkit.org/271758@main">https://commits.webkit.org/271758@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7775752954c276cc1176705913fb87b2f5716c4a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29580 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8241 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30891 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/32096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10392 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5532 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/32096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29853 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/25257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/32096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/33440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/33440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/33440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/7704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3804 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3096,6 +3106,15 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+}
+
+
```
remove extra newlines

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
```diff
@@ -105,6 +105,25 @@
+
+    m_registeredContentScriptsDirectory = FileSystem::pathByAppendingComponent(m_storageDirectory, "RegisteredContentScripts"_s);
+    if (!FileSystem::makeAllDirectories(m_registeredContentScriptsDirectory)) {
+        m_registeredContentScriptsDirectory = String();
```
`nullString()` instead.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
```diff
@@ -105,6 +105,25 @@
     return m_declarativeNetRequestStoreDirectory;
 }
 
+String WebExtensionControllerConfiguration::registeredContentScriptsDirectory()
```
This needs to be elsewhere, since it is per Context. You will want to put this on WebExtensionContext, and use that per-extension subdirectory.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
```diff
@@ -76,6 +75,8 @@ NSArray *toWebAPI(const Vector<WebExtensionScriptInjectionResultParameters>&, bo
 NSArray *toWebAPI(const Vector<WebExtensionRegisteredScriptParameters>&);
 NSString *toWebAPI(WebExtension::InjectionTime);
 
+void parseRegisteredContentScripts(NSArray *, FirstTimeRegistration, Vector<WebExtensionRegisteredScriptParameters>&);
```
This can be a static on the class, so it does not need to move out.

`static void parseRegisteredContentScripts(NSArray *, FirstTimeRegistration, Vector<WebExtensionRegisteredScriptParameters>&);` in the `class` above, like it was before. Might need to make it `public` or add `friend class Foo` for the class that needs to access it. `WebExtensionAPIAction` does this for `WebExtensionAPIMenus`.

xeenon reviewed (changes requested):
Needs changes for the storage location. Otherwise looks good!

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
```diff
@@ -105,6 +105,25 @@
+        return m_registeredContentScriptsDirectory;
+
+    if (!storageIsPersistent()) {
+        m_registeredContentScriptsDirectory = FileSystem::createTemporaryDirectory(@"RegisteredContentScripts");
```
We should use in-memory DBs for this case, so we don't put things on disk. dNR needed on-disk storage to work, so that is why it is an exception.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3096,6 +3105,28 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    }
+
+    return m_registeredContentScriptsDirectory;
+}
```
I don't think we need a sub-directory for one DB, also `ensureDirectoryExists` happens in `_WKWebExtensionSQLiteStore` already so `FileSystem::makeAllDirectories` isn't needed here if we keep this.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3096,6 +3105,28 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+RetainPtr<_WKWebExtensionRegisteredScriptsSQLiteStore> WebExtensionContext::registeredContentScriptsStore()
+{
+    if (!m_registeredContentScriptsStorage)
+        m_registeredContentScriptsStorage = [[_WKWebExtensionRegisteredScriptsSQLiteStore alloc] initWithUniqueIdentifier:m_uniqueIdentifier directory:registeredContentScriptsDirectory() usesInMemoryDatabase:!storageIsPersistent()];
```
This can just use `storageDirectory()`, I don't think we need a sub-directory.

xeenon reviewed (changes requested):
Looking better! Just some minor things to fix.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.mm:
```diff
@@ -0,0 +1,295 @@
+
+    ASSERT(_directory);
+
+    NSString *databaseName = @"registered-content-scripts.db";
```
Can you rename this to "RegisteredContentScripts.db"? I'm working to clean up the files names we use.

kiaraarose committed:
Hook up storage for persistent registered content scripts.
https://bugs.webkit.org/show_bug.cgi?id=265899

Reviewed by Timothy Hatcher.

This is the final work for scripting API. This patch creates a _WKWebExtensionRegisteredScriptsSQLiteStore
which will be used to manage storage for registered scripts. This patch also loads these scripts
from storage when an extension loads.

When an extension updates, we will need to remove these scripts from storage since they shouldn't
persist in this case. This is being tracked in https://bugs.webkit.org/show_bug.cgi?id=249266.

* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.mm: Added.
(rowFilterStringFromRowKeys):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore initWithUniqueIdentifier:directory:usesInMemoryDatabase:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore updateScripts:completionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore deleteScriptsWithIDs:completionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore addScripts:completionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore getScriptsWithCompletionHandler:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _getScriptsWithOutErrorMessage:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _getKeysAndValuesFromRowEnumerator:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _insertScript:inDatabase:errorMessage:]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _currentDatabaseSchemaVersion]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _databaseURL]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _createFreshDatabaseSchema]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _resetDatabaseSchema]):
(-[_WKWebExtensionRegisteredScriptsSQLiteStore _isDatabaseEmpty]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingRegisterContentScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingGetRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterContentScripts):
(WebKit::WebExtensionContext::loadRegisteredContentScripts):
(WebKit::WebExtensionContext::createInjectedContentForScripts):
Renamed from parseRegisteredContentScripts. Parse scripts and create injected content data.
(WebKit::WebExtensionContext::parseRegisteredContentScripts): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::unload):
(WebKit::WebExtensionContext::invalidateStorage):
Invalidate all references to extension storage.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::registeredContentScriptsDirectory):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::parseRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::parseRegisteredContentScripts): Deleted.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:

Canonical link: https://commits.webkit.org/271758@main
